### ISSUE TITLE
Numerical instability with trtllm_fp8_per_tensor_scale_moe

### ISSUE THREAD

**Issue by tantara** [flashinfer-ai/flashinfer#2431]
I'm testing trtllm_fp8_per_tensor_scale_moe with Qwen3 models but seeing NaN. Here are a few data points:

## Qwen3-30B-A3B-Instruct-2507-FP8

It works well with the kernel. I observed latency improvement with neutral accuracy (vs fp8 blockscale). I used `routing_method_type=1`.

## Qwen3-Coder-480B-A35B-Instruct-FP8

Initially, I used the same setting with Qwen3 30B but I saw NaN quite frequently. After I switched routing method to 4, NaN issue was alleviated in case of using shorter context requests (ie <1k tokens). It also shows better accuracy on dataset consists of shorter context compared to routing method 1.

For longer context (ie >10k tokens), NaN values appear again. I changed knobs of `enable_pdl` and `tune_max_num_tokens` but it didn't help that much.

```python
class RoutingMethodType(IntEnum):
    # Default: Softmax -> TopK
    Default = (0,)
    # Renormalize: TopK -> Softmax
    Renormalize = (1,)
    # DeepSeekV3: Sigmoid -> RoutingBiasAdd -> Top2 in group -> Top4 groups -> Top8 experts from the Top4 groups
    DeepSeekV3 = (2,)
    # Llama4: Top1 -> Sigmoid
    Llama4 = (3,)
    # Qwen3: Softmax -> TopK -> Renormalize
    RenormalizeNaive = (4,)
    # TopK only (no softmax)
    TopK = (5,)
    # Unspecified
    Unspecified = 6
```

from https://github.com/flashinfer-ai/flashinfer/blob/48966b603ebabcd003f7272e8dc161ce0575fc9f/flashinfer/fused_moe/core.py#L61

**Comment by djns99** [comment]
I’m not an expert in these kernels, but I notice there is a special case for experts <= 128 [here](https://github.com/flashinfer-ai/flashinfer/blob/48966b603ebabcd003f7272e8dc161ce0575fc9f/include/flashinfer/trtllm/fused_moe/RoutingKernelTopK.cuh#L34) and [here](https://github.com/flashinfer-ai/flashinfer/blob/48966b603ebabcd003f7272e8dc161ce0575fc9f/include/flashinfer/trtllm/fused_moe/RoutingKernelTopK.cuh#L208-L210). Its possible that the >128 path introduces the errors (most likely incorrect padding if I was guessing). This would explain why it works for 128 experts but fails with 160.

**Comment by tantara** [comment]
@djns99 Thanks for your answer. I wonder if other trtllm kernels suffer from numeric instability as long as they go though the same routing code

**Comment by bkryu** [comment]
The issue is stale and should have been fixed by followup TRTLLM MoE cubin updates, including #2803

Issue does not reproduce with the following script

```python
import sys
import traceback

import torch

import tests.moe.test_trtllm_gen_fused_moe as T  # noqa: E402
from flashinfer.fused_moe import RoutingMethodType  # noqa: E402
from flashinfer import fp4_quantize  # noqa: F401,E402

HIDDEN = 1024
INTERMEDIATE = 1024
NUM_EXPERTS = 160  # Qwen3-Coder-480B-A35B -> the >128 boundary case from the issue
TOP_K = 8

# Build the cache_permute_indices dict the harness expects (it's a module-scoped
# pytest fixture; we just emulate it).
cache_permute_indices = {}

results = []


def run_one(routing_method_type, num_tokens, rm_name):
    cfg = {
        "num_experts": NUM_EXPERTS,
        "top_k": TOP_K,
        "padding": 8,
        "n_groups": None,
        "top_k_groups": None,
        "routed_scaling": None,
        "has_routing_bias": False,
        "routing_method_type": routing_method_type,
        "compatible_moe_impls": [T.FP8PerTensorMoe],
        "compatible_intermediate_size": [INTERMEDIATE],
        "enable_autotune": True,
    }
    weight_processing = {
        "use_shuffled_weight": True,
        "layout": T.WeightLayout.MajorK,
        "compatible_moe_impls": [T.FP8PerTensorMoe],
    }
    moe_impl = T.FP8PerTensorMoe()
    tag = f"routing={rm_name}({int(routing_method_type)}) tokens={num_tokens} experts={NUM_EXPERTS} topk={TOP_K}"
    print(f"\n==== RUN {tag} ====", flush=True)
    try:
        T.run_moe_test(
            num_tokens=num_tokens,
            hidden_size=HIDDEN,
            intermediate_size=INTERMEDIATE,
            moe_impl=moe_impl,
            routing_config=cfg,
            weight_processing=weight_processing,
            activation_type=T.ActivationType.Swiglu.value,
            cache_permute_indices=cache_permute_indices,
            routing_logits_dtype=torch.bfloat16,
            zero_hidden_states=False,
        )
        print(f"RESULT {tag}: PASS (no NaN, accuracy within tolerance)", flush=True)
        results.append((tag, "PASS"))
    except Exception as e:
        msg = f"{type(e).__name__}: {e}"
        is_nan = "nan" in msg.lower() or "inf" in msg.lower()
        print(f"RESULT {tag}: FAIL -> {msg}", flush=True)
        traceback.print_exc()
        results.append((tag, f"FAIL({'NaN' if is_nan else 'other'}): {msg}"))


def main():
    print("torch", torch.__version__, "device", torch.cuda.get_device_name(0), flush=True)
    print("capability", torch.cuda.get_device_capability(0), flush=True)
    # routing_method_type 1 = Renormalize, 4 = RenormalizeNaive (from the issue)
    for rm, name in [
        (RoutingMethodType.Renormalize, "Renormalize"),
        (RoutingMethodType.RenormalizeNaive, "RenormalizeNaive"),
    ]:
        for nt in [8, 3072]:  # short context and long-ish context
            run_one(rm, nt, name)

    print("\n==== SUMMARY ====", flush=True)
    any_nan = False
    for tag, r in results:
        print(f"  {tag}: {r}", flush=True)
        if "NaN" in r:
            any_nan = True
    print(f"\nANY_NAN={any_nan}", flush=True)
    print("OVERALL=" + ("REPRODUCED_NAN" if any_nan else "NO_NAN"), flush=True)


if __name__ == "__main__":
    main()

```

Closing but @tantara and @djns99 , please feel free to reopen if the issue persists
